File: src/main/java/ai/ancf/lmos/operator/resolver/impl/AgentResolver.java
/*
 * SPDX-FileCopyrightText: 2024 Deutsche Telekom AG
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package ai.ancf.lmos.operator.resolver.impl;

import ai.ancf.lmos.operator.resolver.*;
import ai.ancf.lmos.operator.resources.agent.AgentResource;
import ai.ancf.lmos.operator.resources.channel.RequiredCapability;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashSet;
import java.util.Set;

public class AgentResolver implements Resolver<AgentResource> {

    private static final Logger LOG = LoggerFactory.getLogger(AgentResolver.class);

    @Override
    @NotNull
    public Set<Wire<AgentResource>> resolve(Set<RequiredCapability> requiredCapabilities, ResolveContext<AgentResource> resolveContext) throws ResolverException {

        var agentResources = resolveContext.getAvailableResources();

        if(agentResources == null || agentResources.isEmpty()){
            throw new ResolverException("Resolve context is empty", requiredCapabilities);
        }

        var bestMatchingCapabilities = new HashSet<Wire<AgentResource>>();
        var unresolvedRequiredCapabilities = new HashSet<RequiredCapability>();
        requiredCapabilities.forEach(
                requiredCapability -> {
                    Set<Wire<AgentResource>> wireCapabilities = resolveContext.findCapabilityProviders(requiredCapability);
                    if(wireCapabilities.isEmpty()){
                        unresolvedRequiredCapabilities.add(requiredCapability);
                    }else{
                        ResolveStrategy strategy = requiredCapability.getStrategy();
                        var bestMatchingWire = CapabilityResolver.findBestMatchingWire(wireCapabilities, strategy);
                        if(bestMatchingWire.isPresent()){
                            bestMatchingCapabilities.add(bestMatchingWire.get());
                        }else{
                            unresolvedRequiredCapabilities.add(requiredCapability);
                        }
                    }
                }
        );
        if(!unresolvedRequiredCapabilities.isEmpty()){
            throw new ResolverException("Required capabilities not resolved", unresolvedRequiredCapabilities);
        }

        return bestMatchingCapabilities;
    }

}

File: src/main/kotlin/org/eclipse/lmos/operator/resolver/impl/AgentResolver.kt
/*
 * SPDX-FileCopyrightText: 2024 Deutsche Telekom AG
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package org.eclipse.lmos.operator.resolver.impl

import org.eclipse.lmos.operator.resolver.ResolveContext
import org.eclipse.lmos.operator.resolver.Resolver
import org.eclipse.lmos.operator.resolver.ResolverException
import org.eclipse.lmos.operator.resolver.Wire
import org.eclipse.lmos.operator.resources.AgentResource
import org.eclipse.lmos.operator.resources.RequiredCapability
import java.util.function.Consumer

class AgentResolver : Resolver<AgentResource> {
    @Throws(ResolverException::class)
    override fun resolve(
        requiredCapabilities: Set<RequiredCapability>,
        resolveContext: ResolveContext<AgentResource>,
    ): Set<Wire<AgentResource>> {
        val agentResources = resolveContext.availableResources

        if (agentResources.isEmpty()) {
            throw ResolverException("Resolve context is empty", requiredCapabilities)
        }

        val bestMatchingCapabilities = mutableSetOf<Wire<AgentResource>>()
        val unresolvedRequiredCapabilities = mutableSetOf<RequiredCapability>()
        requiredCapabilities.forEach(
            Consumer { requiredCapability: RequiredCapability ->
                val wireCapabilities = resolveContext.findCapabilityProviders(requiredCapability)
                if (wireCapabilities.isEmpty()) {
                    unresolvedRequiredCapabilities.add(requiredCapability)
                } else {
                    val strategy = requiredCapability.strategy
                    val bestMatchingWire =
                        CapabilityResolver.findBestMatchingWire(
                            wireCapabilities,
                            strategy,
                        )
                    if (bestMatchingWire != null) {
                        bestMatchingCapabilities.add(bestMatchingWire)
                    } else {
                        unresolvedRequiredCapabilities.add(requiredCapability)
                    }
                }
            },
        )
        if (unresolvedRequiredCapabilities.isNotEmpty()) {
            throw ResolverException("Required capabilities not resolved", unresolvedRequiredCapabilities)
        }

        return bestMatchingCapabilities
    }
}
